scripts/hud/customizer.ts

```diff
@@ -0,0 +1,654 @@
+import { values } from 'util/enum';
+import { PanelHandler, OnPanelLoad } from 'util/module-helpers';
+import * as LayoutUtil from 'common/layout';
+
+enum Axis {
+	X = 0,
+	Y = 1
+}
+
+const Axes = [Axis.X, Axis.Y] as const;
+
+type Snaps = [SnapMode, SnapMode];
+
+/**
+ * The four different snapping behaviours
+ *
+ * For a horizontal panel, min and max are left and right respectively, for vertical, top and bottom
+ */
+enum SnapMode {
+	MIN = 1,
+	MID = 2,
+	MAX = 3,
+	OFF = 4
+}
+
+// Tuple of X, Y axis snaps respectively
+const DEFAULT_SNAP_MODES: Snaps = [SnapMode.OFF, SnapMode.OFF];
+
+enum ResizeDirection {
+	TOP = 1,
+	TOP_RIGHT = 2,
+	RIGHT = 3,
+	BOTTOM_RIGHT = 4,
+	BOTTOM = 5,
+	BOTTOM_LEFT = 6,
+	LEFT = 7,
+	TOP_LEFT = 8
+}
+
+enum DragMode {
+	MOVE = 0,
+	RESIZE_TOP = 1,
+	RESIZE_TOP_RIGHT = 2,
+	RESIZE_RIGHT = 3,
+	RESIZE_BOTTOM_RIGHT = 4,
+	RESIZE_BOTTOM = 5,
+	RESIZE_BOTTOM_LEFT = 6,
+	RESIZE_LEFT = 7,
+	RESIZE_TOP_LEFT = 8
+}
+
+const resizeVector: ReadonlyMap<Exclude<DragMode, DragMode.MOVE>, [number, number]> = new Map([
+	[DragMode.RESIZE_TOP, [0, -1]],
+	[DragMode.RESIZE_TOP_RIGHT, [1, -1]],
+	[DragMode.RESIZE_RIGHT, [1, 0]],
+	[DragMode.RESIZE_BOTTOM_RIGHT, [1, 1]],
+	[DragMode.RESIZE_BOTTOM, [0, 1]],
+	[DragMode.RESIZE_BOTTOM_LEFT, [-1, 1]],
+	[DragMode.RESIZE_LEFT, [-1, 0]],
+	[DragMode.RESIZE_TOP_LEFT, [-1, -1]]
+]);
+
+const MAX_X_POS = 1920;
+const MAX_Y_POS = 1080;
+const MAX_OOB = 16;
+
+interface HudConfig {
+	components: {
+		[id: string]: {
+			position: LayoutUtil.Position;
+			size: LayoutUtil.Size;
+			snaps: [number, number];
+		};
+	};
+	settings: {
+		gridSize: number;
+	};
+}
+
+const DEFAULT_GRID_SIZE = 5;
+
+interface Component {
+	panel: GenericPanel;
+	snaps: Snaps;
+	position: LayoutUtil.Position;
+	size: LayoutUtil.Size;
+	properties?: Record<string, unknown>;
+}
+
+interface Gridline {
+	panel: Panel;
+	offset: number;
+}
+
+type GridlineForAxis = [Gridline[], Gridline[]];
+
+const snaps: Record<
```

Comment:
Seeing as this is never mutated I'd name this `SNAPS`.

Reply:
Addressed in 74f396aa55355f1288479bab2290b85024d24fa6